src/codec/mod.rs

```diff
@@ -0,0 +1,210 @@
+//! A collection of codecs that can be used to transform between bytes streams /
+//! byte messages, byte frames and structured events.
+
+#![deny(missing_docs)]
+
+mod framers;
+mod parsers;
+
+use crate::{
+    event::Event,
+    internal_events::{DecoderFramingFailed, DecoderParseFailed},
+};
+use bytes::{Bytes, BytesMut};
+pub use framers::*;
+pub use parsers::*;
+use serde::{Deserialize, Serialize};
+use smallvec::SmallVec;
+
+/// An error that occurred while decoding structured events from a byte stream /
+/// byte messages.
+#[derive(Debug)]
+pub enum Error {
+    /// The error occurred while producing byte frames from the byte stream /
+    /// byte messages.
+    FramingError(BoxedFramingError),
+    /// The error occurred while parsing structured events from a byte frame.
+    ParsingError(crate::Error),
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::FramingError(error) => write!(formatter, "FramingError({})", error),
+            Self::ParsingError(error) => write!(formatter, "ParsingError({})", error),
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+impl From<std::io::Error> for Error {
+    fn from(error: std::io::Error) -> Self {
+        Self::FramingError(Box::new(error))
+    }
+}
+
+#[derive(Clone)]
+/// A decoder that can decode structured events from a byte stream / byte
+/// messages.
+pub struct Decoder {
+    framer: BoxedFramer,
+    parser: BoxedParser,
+}
+
+impl Decoder {
+    /// Creates a new `Decoder` with the specified `Framer` to produce byte
+    /// frames from the byte stream / byte messages and `Parser` to parse
+    /// structured events from a byte frame.
+    pub fn new(framer: BoxedFramer, parser: BoxedParser) -> Self {
+        Self { framer, parser }
+    }
+
+    /// Method to combine framing and parsing, such that an incoming byte stream
+    /// / byte messages are transformed directly to structured events.
+    ///
+    /// Zero-byte frames are skipped without parsing.
+    fn decode(
+        &mut self,
+        buf: &mut BytesMut,
+        decode_frame: impl Fn(
+            &mut BoxedFramer,
+            &mut BytesMut,
+        ) -> Result<Option<Bytes>, BoxedFramingError>,
+    ) -> Result<Option<(SmallVec<[Event; 1]>, usize)>, Error> {
+        loop {
+            // Frame bytes from the incoming byte stream / byte messages.
+            let frame = decode_frame(&mut self.framer, buf).map_err(|error| {
+                emit!(DecoderFramingFailed { error: &error });
+                Error::FramingError(error)
+            })?;
+
+            break if let Some(frame) = frame {
+                let byte_size = frame.len();
+
+                // Skip zero-sized frames.
+                if byte_size == 0 {
+                    continue;
+                }
```

Comment:
I think I introduced this when handling multiple (empty) newlines in NDJSON. I agree that the parser should have an option to ignore empty frames, rather than doing this unconditionally in the framer